radicle-tracker/src/ops/sequence.rs

```diff
@@ -0,0 +1,361 @@
+    where
+        T: Apply<Op = M> + Ord,
+        O: Ord + Clone,
+    {
+        match op {
+            Op::Append { id, val, .. } => match self.find_item(&id) {
+                Err(ix) => {
+                    self.val.insert(ix, (id, val));
+                    Ok(())
+                },
+                Ok(ix) => {
+                    // ix exists because of find_item. qed.
+                    let item = &self.val[ix];
+                    if item.0 < id {
+                        self.val.insert(ix + 1, (id, val));
+                    } else {
+                        self.val.insert(ix, (id, val));
+                    }
+
+                    Ok(())
+                },
+            },
+            Op::Modify { id, ix, op } => match self.find_item(&id) {
+                Ok(ix) => {
+                    // ix exists because of find_item. qed.
+                    let (_, item) = &mut self.val[ix];
+                    item.apply(op)?;
+                    Ok(())
+                },
+                Err(_) => Err(Error::IndexOutOfBounds(ix)),
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::ops::id::UniqueTimestamp;
+    use std::{cmp::Ordering, convert::Infallible, error};
+
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    struct Int {
+        id: usize,
+        val: u32,
+    }
+
+    impl Int {
+        fn new(i: u32) -> Self {
+            Int {
+                id: i as usize,
+                val: i,
+            }
+        }
+    }
+
+    impl PartialOrd for Int {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            self.id.partial_cmp(&other.id)
+        }
+    }
+
+    impl Ord for Int {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.id.cmp(&other.id)
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+    struct Add(u32);
+
+    impl Apply for Int {
+        type Op = Add;
+        type Error = Infallible;
+
+        fn apply(&mut self, op: Self::Op) -> Result<(), Self::Error> {
+            self.val += op.0;
+            Ok(())
+        }
+    }
+
+    type TestResult = Result<(), Box<dyn error::Error + 'static>>;
+
+    #[test]
+    fn sync_appends() -> TestResult {
+        let mut left: OrdSequence<UniqueTimestamp, Int> = OrdSequence::new();
+        let append1 = left.append(Int::new(1));
+        let append2 = left.append(Int::new(2));
+
+        let mut right = OrdSequence::new();
+        right.apply(append1)?;
+        right.apply(append2)?;
+
+        assert_eq!(left, right);
+
+        Ok(())
+    }
+
+    #[test]
+    fn out_of_order_appends_fail() -> TestResult {
+        let mut left: OrdSequence<UniqueTimestamp, Int> = OrdSequence::new();
+        let append1 = left.append(Int::new(1));
+        let append2 = left.append(Int::new(2));
+
+        let mut right = OrdSequence::new();
+        right.apply(append2)?;
+        right.apply(append1)?;
```

Comment:
So I accidentally realised that this commutes due inserting based on the timestamp. I dunno if this is desireable or if we'd rather insert by the index and _fall back_ to the timestamp.